From 8a2d9c5d3c84a8edc7faebd60bd22ef5f5bd27d1 Mon Sep 17 00:00:00 2001
From: Amit Kumar Prasad <quic_amitpras@quicinc.com>
Date: Fri, 8 Jul 2022 16:31:59 +0530
Subject: [PATCH 041/100] A2DP Sink: Add support for AAC codec in Split mode

- AAC codec is added for split a2dp sink feature.

- AAC codec always enabled for all source devices,
  no whitelist/blacklist mechanism added for AAC
  codec in sink role.

- Changed the COD to Audio wearable device for
  Aurora.

CRs-Fixed: 3245176
Change-Id: I601ad6e4a7197ee773575e2095ae7381d1ed472c
---
 bta/dm/bta_dm_act.cc            |  13 ++-
 btif/co/bta_av_co.cc            |  46 ++++++++---
 stack/a2dp/a2dp_aac.cc          | 141 ++++++++++++++++++++++++++++++--
 stack/a2dp/a2dp_codec_config.cc |  77 ++++++++++++++++-
 stack/a2dp/a2dp_sbc.cc          |  13 ++-
 stack/a2dp/a2dp_vendor.cc       |   8 ++
 stack/avdt/avdt_ccb_act.cc      |   9 +-
 stack/include/a2dp_aac.h        |  37 +++++++++
 stack/include/a2dp_codec_api.h  |  13 +++
 stack/include/a2dp_sbc.h        |   5 ++
 10 files changed, 341 insertions(+), 21 deletions(-)
 mode change 100755 => 100644 stack/a2dp/a2dp_aac.cc

diff --git a/bta/dm/bta_dm_act.cc b/bta/dm/bta_dm_act.cc
index 1ae0b8374..19b76ecb1 100644
--- a/bta/dm/bta_dm_act.cc
+++ b/bta/dm/bta_dm_act.cc
@@ -427,6 +427,7 @@ static void bta_dm_sys_hw_cback(tBTA_SYS_HW_EVT status) {
   uint8_t key_mask = 0;
   tBTA_BLE_LOCAL_ID_KEYS id_key;
   tBTA_DM_MSG* p_data;
+  char board_name[PROPERTY_VALUE_MAX];
 #ifdef ADV_AUDIO_FEATURE
   char adv_audio_support_prop_value[PROPERTY_VALUE_MAX];
 #endif
@@ -525,8 +526,14 @@ static void bta_dm_sys_hw_cback(tBTA_SYS_HW_EVT status) {
                        dev_class[0], dev_class[1], dev_class[2]);
     }
 #endif
-
-    BTM_SetDeviceClass(dev_class);
+    osi_property_get("ro.board.platform", board_name, "");
+    if (!strncmp("neo", board_name, 3)) {
+      APPL_TRACE_DEBUG("%s chanding class name for neo", __func__);
+      dev_class[0] = 0x14; // minor dev class as Glass
+      dev_class[1] = 0x07; // major dev class as Wearable
+      dev_class[1] = 0x20; // Service class as Audio
+      BTM_SetDeviceClass(dev_class);
+    }
 
     /* load BLE local information: ID keys, ER if available */
     Octet16 er;
@@ -1014,7 +1021,7 @@ void bta_dm_remove_device(tBTA_DM_MSG* p_data) {
     }
    /*
     The remote device is not in bta_dm_cb.device_list because remote extended
-    features complete event is not retured.   The reason may be remote device 
+    features complete event is not retured.   The reason may be remote device
     is in bad state and has no response.
     For this case, ACL link shall be removed also.
     */
diff --git a/btif/co/bta_av_co.cc b/btif/co/bta_av_co.cc
index 76a52927d..c6ec7f2db 100644
--- a/btif/co/bta_av_co.cc
+++ b/btif/co/bta_av_co.cc
@@ -49,6 +49,11 @@
  *
  ******************************************************************************/
 
+/*
+ * Changes from Qualcomm Innovation Center are provided under the following license:
+ * Copyright (c) 2022 Qualcomm Innovation Center, Inc. All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
+ */
 
 /******************************************************************************
  *
@@ -174,6 +179,7 @@ extern bool btif_av_check_is_reconfig_pending_flag_set(RawAddress address);
 extern bool btif_av_check_is_cached_reconfig_event_exist(RawAddress address);
 extern bool btif_av_check_is_retry_reconfig_set(RawAddress address);
 extern void btif_av_clear_is_retry_reconfig_flag(RawAddress address);
+extern bool is_a2dp_split_sink_enabled();
 
 /*******************************************************************************
  **
@@ -768,7 +774,7 @@ void bta_av_co_audio_setconfig(tBTA_AV_HNDL hndl, const uint8_t* p_codec_info,
 
     if (t_local_sep == AVDT_TSEP_SNK) {
       APPL_TRACE_DEBUG("%s: peer is A2DP SRC", __func__);
-      codec_config_supported = A2DP_IsSinkCodecSupported(p_codec_info);
+      codec_config_supported = A2DP_IsPeerSourceCodecSupported(p_codec_info);
       if (codec_config_supported) {
         // If Peer is SRC, and our config subset matches with what is
         // requested by peer, then just accept what peer wants.
@@ -1213,12 +1219,21 @@ static const tBTA_AV_CO_SINK* bta_av_co_find_peer_src_supports_codec(
     const tBTA_AV_CO_PEER* p_peer) {
   APPL_TRACE_DEBUG("%s: peer num_sup_srcs = %d", __func__,
                    p_peer->num_sup_srcs);
-
-  for (size_t index = 0; index < p_peer->num_sup_srcs; index++) {
-    const uint8_t* p_codec_caps = p_peer->srcs[index].codec_caps;
-    if (A2DP_CodecTypeEquals(bta_av_co_cb.codec_config, p_codec_caps) &&
-        A2DP_IsPeerSourceCodecSupported(p_codec_caps)) {
-      return &p_peer->srcs[index];
+  /*Select Peer Source supported codec as per local Sink order*/
+  for (const auto& iter : p_peer->codecs->orderedSinkCodecs()) {
+    APPL_TRACE_DEBUG("%s: trying codec %s", __func__, iter->name().c_str());
+    for (size_t index = 0; index < p_peer->num_sup_srcs; index++) {
+      const uint8_t* p_codec_caps = p_peer->srcs[index].codec_caps;
+      btav_a2dp_codec_index_t peer_codec_index =
+      A2DP_SinkCodecIndex(p_peer->srcs[index].codec_caps);
+        APPL_TRACE_DEBUG("%s ind: %d, peer_codec_index : %d :: codec_config.codecIndex() : %d",
+                 __func__, index, peer_codec_index, iter->codecIndex());
+      if (peer_codec_index != iter->codecIndex()) {
+        continue;
+      }
+      if (A2DP_IsPeerSourceCodecSupported(p_codec_caps)) {
+        return &p_peer->srcs[index];
+      }
     }
   }
   return NULL;
@@ -1248,7 +1263,11 @@ static bool bta_av_co_check_peer_eligible_for_aac_codec(
     APPL_TRACE_DEBUG("%s: vendor: 0x%04x product: 0x%04x version: 0x%04x", __func__, vendor, product, version);
     vndr_prdt_ver_present = true;
   }
-  if (vndr_prdt_ver_present && (vendor == A2DP_AAC_BOSE_VENDOR_ID)) {
+
+  if (is_a2dp_split_sink_enabled()) {
+    APPL_TRACE_DEBUG("%s: peer is A2DP SRC", __func__);
+    aac_support = true;
+  } else if (vndr_prdt_ver_present && (vendor == A2DP_AAC_BOSE_VENDOR_ID)) {
     APPL_TRACE_DEBUG("%s: vendor id info matches ", __func__);
     vndr_prdt_ver_present = false;
     aac_support = true;
@@ -2173,6 +2192,7 @@ void bta_av_co_init(
   RawAddress bt_addr;
   const char *a2dp_ofload_cap = controller_get_interface()->get_a2dp_offload_cap();
   tBTA_AV_CO_PEER* p_peer;
+  char value[PROPERTY_VALUE_MAX] = {'\0'};
   /* Protect access to bta_av_co_cb.codec_config */
   mutex_global_lock();
   /* Reset the control block */
@@ -2184,13 +2204,19 @@ void bta_av_co_init(
   bta_av_co_cp_set_flag(AVDT_CP_SCMS_COPY_FREE);
 #endif
 
+  /*A2DP Sink codecs*/
+  osi_property_get("persist.vendor.bluetooth.a2dp_sink_cap", value, "sbc");
+  A2DP_SetSinkCodec(value);
+  /*A2DP Sink codecs*/
+
 /* SPLITA2DP */
   bool a2dp_offload = btif_av_is_split_a2dp_enabled();
   bool isScramblingSupported = bta_av_co_is_scrambling_enabled();
   bool is44p1kFreqSupported = bta_av_co_is_44p1kFreq_enabled();
-
-  A2DP_SetOffloadStatus(a2dp_offload, a2dp_ofload_cap, isScramblingSupported,
+  if(!is_a2dp_split_sink_enabled()) {
+    A2DP_SetOffloadStatus(a2dp_offload, a2dp_ofload_cap, isScramblingSupported,
                        is44p1kFreqSupported, offload_enabled_codecs_config);
+  }
 
 /* SPLITA2DP */
   bool isMcastSupported = btif_av_is_multicast_supported();
diff --git a/stack/a2dp/a2dp_aac.cc b/stack/a2dp/a2dp_aac.cc
old mode 100755
new mode 100644
index 3b300fcd7..0a83b60d5
--- a/stack/a2dp/a2dp_aac.cc
+++ b/stack/a2dp/a2dp_aac.cc
@@ -14,6 +14,12 @@
  * limitations under the License.
  */
 
+/*
+ * Changes from Qualcomm Innovation Center are provided under the following license:
+ * Copyright (c) 2022 Qualcomm Innovation Center, Inc. All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
+ */
+
 /******************************************************************************
  *
  *  Utility functions to help build and parse the AAC Codec Information
@@ -56,6 +62,21 @@ static const tA2DP_AAC_CIE a2dp_aac_src_caps = {
     // bits_per_sample
     BTAV_A2DP_CODEC_BITS_PER_SAMPLE_16};
 
+/* AAC Sink codec capabilities */
+static const tA2DP_AAC_CIE a2dp_aac_sink_caps = {
+    // objectType
+    A2DP_AAC_OBJECT_TYPE_MPEG2_LC,
+    // sampleRate
+    A2DP_AAC_SAMPLING_FREQ_44100 | A2DP_AAC_SAMPLING_FREQ_48000,
+    // channelMode
+    A2DP_AAC_CHANNEL_MODE_MONO | A2DP_AAC_CHANNEL_MODE_STEREO,
+    // variableBitRateSupport
+    A2DP_AAC_VARIABLE_BIT_RATE_ENABLED,
+    // bitRate
+    A2DP_AAC_DEFAULT_BITRATE,
+    // bits_per_sample
+    BTAV_A2DP_CODEC_BITS_PER_SAMPLE_16};
+
 static const tA2DP_AAC_CIE a2dp_aac_src_vbr_caps = {
     // objectType
     A2DP_AAC_OBJECT_TYPE_MPEG2_LC,
@@ -308,11 +329,19 @@ bool A2DP_IsSourceCodecValidAac(const uint8_t* p_codec_info) {
 }
 
 bool A2DP_IsSinkCodecValidAac(UNUSED_ATTR const uint8_t* p_codec_info) {
-  return false;
+    tA2DP_AAC_CIE cfg_cie;
+
+    /* Use a liberal check when parsing the codec info */
+    return (A2DP_ParseInfoAac(&cfg_cie, p_codec_info, false) == A2DP_SUCCESS) ||
+           (A2DP_ParseInfoAac(&cfg_cie, p_codec_info, true) == A2DP_SUCCESS);
 }
 
 bool A2DP_IsPeerSourceCodecValidAac(UNUSED_ATTR const uint8_t* p_codec_info) {
-  return false;
+	  tA2DP_AAC_CIE cfg_cie;
+
+  /* Use a liberal check when parsing the codec info */
+  return (A2DP_ParseInfoAac(&cfg_cie, p_codec_info, false) == A2DP_SUCCESS) ||
+         (A2DP_ParseInfoAac(&cfg_cie, p_codec_info, true) == A2DP_SUCCESS);
 }
 
 bool A2DP_IsPeerSinkCodecValidAac(const uint8_t* p_codec_info) {
@@ -330,17 +359,46 @@ bool A2DP_GetAacCIE(const uint8_t* p_codec_info,
 }
 
 bool A2DP_IsSinkCodecSupportedAac(UNUSED_ATTR const uint8_t* p_codec_info) {
-  return false;
+  return A2DP_CodecInfoMatchesCapabilityAac(&a2dp_aac_sink_caps, p_codec_info,
+                                            false) == A2DP_SUCCESS;
 }
 
 bool A2DP_IsPeerSourceCodecSupportedAac(
     UNUSED_ATTR const uint8_t* p_codec_info) {
-  return false;
+  return A2DP_CodecInfoMatchesCapabilityAac(&a2dp_aac_sink_caps, p_codec_info,
+                                            true) == A2DP_SUCCESS;
 }
 
 tA2DP_STATUS A2DP_BuildSrc2SinkConfigAac(UNUSED_ATTR const uint8_t* p_src_cap,
                                          UNUSED_ATTR uint8_t* p_pref_cfg) {
-  return A2DP_NS_CODEC_TYPE;
+  tA2DP_AAC_CIE src_cap;
+  tA2DP_AAC_CIE pref_cap;
+
+  /* initialize it to default AAC configuration */
+  A2DP_BuildInfoAac(AVDT_MEDIA_TYPE_AUDIO, &a2dp_aac_default_config,
+                    p_pref_cfg);
+
+  /* now try to build a preferred one */
+  /* parse configuration */
+  tA2DP_STATUS status = A2DP_ParseInfoAac(&src_cap, p_src_cap, true);
+  if (status != A2DP_SUCCESS) {
+    LOG_ERROR(LOG_TAG, "%s: can't parse src cap ret = %d", __func__, status);
+    return A2DP_FAIL;
+  }
+
+  if (src_cap.sampleRate & A2DP_AAC_SAMPLING_FREQ_48000)
+    pref_cap.sampleRate = A2DP_AAC_SAMPLING_FREQ_48000;
+  else if (src_cap.sampleRate & A2DP_AAC_SAMPLING_FREQ_44100)
+    pref_cap.sampleRate = A2DP_AAC_SAMPLING_FREQ_44100;
+
+  if (src_cap.channelMode & A2DP_AAC_CHANNEL_MODE_STEREO)
+    pref_cap.channelMode = A2DP_AAC_CHANNEL_MODE_STEREO;
+  else if (src_cap.channelMode & A2DP_AAC_CHANNEL_MODE_MONO)
+    pref_cap.channelMode = A2DP_AAC_CHANNEL_MODE_MONO;
+
+  A2DP_BuildInfoAac(AVDT_MEDIA_TYPE_AUDIO, &pref_cap, p_pref_cfg);
+
+  return A2DP_SUCCESS;
 }
 
 // Checks whether A2DP AAC codec configuration matches with a device's codec
@@ -514,6 +572,22 @@ int A2DP_GetTrackChannelCountAac(const uint8_t* p_codec_info) {
 }
 
 int A2DP_GetSinkTrackChannelTypeAac(UNUSED_ATTR const uint8_t* p_codec_info) {
+  tA2DP_AAC_CIE aac_cie;
+
+  // Check whether the codec info contains valid data
+  tA2DP_STATUS a2dp_status = A2DP_ParseInfoAac(&aac_cie, p_codec_info, false);
+  if (a2dp_status != A2DP_SUCCESS) {
+    LOG_ERROR(LOG_TAG, "%s: cannot decode codec information: %d", __func__,
+              a2dp_status);
+    return -1;
+  }
+
+  switch (aac_cie.channelMode) {
+    case A2DP_AAC_CHANNEL_MODE_MONO:
+      return 1;
+    case A2DP_AAC_CHANNEL_MODE_STEREO:
+      return 3;
+  }
   return -1;
 }
 
@@ -754,8 +828,15 @@ btav_a2dp_codec_index_t A2DP_SourceCodecIndexAac(
   return BTAV_A2DP_CODEC_INDEX_SOURCE_AAC;
 }
 
+btav_a2dp_codec_index_t A2DP_SinkCodecIndexAac(
+    UNUSED_ATTR const uint8_t* p_codec_info) {
+  return BTAV_A2DP_CODEC_INDEX_SINK_AAC;
+}
+
 const char* A2DP_CodecIndexStrAac(void) { return "AAC"; }
 
+const char* A2DP_CodecIndexStrAacSink(void) { return "AAC SINK"; }
+
 bool A2DP_InitCodecConfigAac(tAVDT_CFG* p_cfg) {
   if (!A2DP_IsCodecEnabled(BTAV_A2DP_CODEC_INDEX_SOURCE_AAC)){
     LOG_ERROR(LOG_TAG, "%s: AAC disabled in both SW and HW mode", __func__);
@@ -778,6 +859,11 @@ bool A2DP_InitCodecConfigAac(tAVDT_CFG* p_cfg) {
   return true;
 }
 
+bool A2DP_InitCodecConfigAacSink(tAVDT_CFG* p_cfg) {
+  return A2DP_BuildInfoAac(AVDT_MEDIA_TYPE_AUDIO, &a2dp_aac_sink_caps,
+                           p_cfg->codec_info) == A2DP_SUCCESS;
+}
+
 UNUSED_ATTR static void build_codec_config(const tA2DP_AAC_CIE& config_cie,
                                            btav_a2dp_codec_config_t* result) {
   if (config_cie.sampleRate & A2DP_AAC_SAMPLING_FREQ_44100)
@@ -1451,3 +1537,48 @@ fail:
          sizeof(ota_codec_peer_config_));
   return false;
 }
+
+A2dpCodecConfigAacSink::A2dpCodecConfigAacSink(
+    btav_a2dp_codec_priority_t codec_priority)
+    : A2dpCodecConfig(BTAV_A2DP_CODEC_INDEX_SINK_AAC,
+                      A2DP_CodecIndexStrAacSink(), codec_priority) {}
+
+A2dpCodecConfigAacSink::~A2dpCodecConfigAacSink() {}
+
+bool A2dpCodecConfigAacSink::init() {
+  if (!isValid()) return false;
+
+  if (A2DP_IsCodecEnabledInSink(BTAV_A2DP_CODEC_INDEX_SINK_AAC)){
+    LOG_DEBUG(LOG_TAG, "%s: AAC is enabled in Sink", __func__);
+    return true;
+  } else {
+    LOG_DEBUG(LOG_TAG, "%s: AAC is disabled in Sink", __func__);
+    return false;
+  }
+}
+
+period_ms_t A2dpCodecConfigAacSink::encoderIntervalMs() const {
+  // TODO: This method applies only to Source codecs
+  return 0;
+}
+
+bool A2dpCodecConfigAacSink::setCodecConfig(
+    UNUSED_ATTR const uint8_t* p_peer_codec_info,
+    UNUSED_ATTR bool is_capability,
+    UNUSED_ATTR uint8_t* p_result_codec_config) {
+  // TODO: This method applies only to Source codecs
+  return false;
+}
+
+bool A2dpCodecConfigAacSink::useRtpHeaderMarkerBit() const {
+  // TODO: This method applies only to Source codecs
+  return false;
+}
+
+bool A2dpCodecConfigAacSink::updateEncoderUserConfig(
+    UNUSED_ATTR const tA2DP_ENCODER_INIT_PEER_PARAMS* p_peer_params,
+    UNUSED_ATTR bool* p_restart_input, UNUSED_ATTR bool* p_restart_output,
+    UNUSED_ATTR bool* p_config_updated) {
+  // TODO: This method applies only to Source codecs
+  return false;
+}
\ No newline at end of file
diff --git a/stack/a2dp/a2dp_codec_config.cc b/stack/a2dp/a2dp_codec_config.cc
index 321312320..d7ce6ad3c 100644
--- a/stack/a2dp/a2dp_codec_config.cc
+++ b/stack/a2dp/a2dp_codec_config.cc
@@ -47,6 +47,12 @@
  * limitations under the License.
  */
 
+/*
+ * Changes from Qualcomm Innovation Center are provided under the following license:
+ * Copyright (c) 2022 Qualcomm Innovation Center, Inc. All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
+ */
+
 /**
  * A2DP Codecs Configuration
  */
@@ -75,6 +81,7 @@
 #include "device/include/device_iot_config.h"
 #include "bta/av/bta_av_int.h"
 #include "btif_av.h"
+#include <hardware/bt_av.h>
 
 /* The Media Type offset within the codec info byte array */
 #define A2DP_MEDIA_TYPE_OFFSET 1
@@ -105,6 +112,9 @@ bool vbr_supported = false;
 bool aptxadaptiver2_1_supported = false;
 bool aptxadaptiver2_2_supported = false;
 std::string offload_caps = "";
+bool sbc_sink = false;
+bool aac_sink = false;
+
 static void init_btav_a2dp_codec_config(
     btav_a2dp_codec_config_t* codec_config, btav_a2dp_codec_index_t codec_index,
     btav_a2dp_codec_priority_t codec_priority) {
@@ -198,6 +208,9 @@ A2dpCodecConfig* A2dpCodecConfig::createCodec(
     case BTAV_A2DP_CODEC_INDEX_SOURCE_APTX_TWS:
       codec_config = new A2dpCodecConfigAptxTWS(codec_priority);
       break;
+    case BTAV_A2DP_CODEC_INDEX_SINK_AAC:
+      codec_config = new A2dpCodecConfigAacSink(codec_priority);
+      break;
     // Add a switch statement for each vendor-specific codec
     case BTAV_A2DP_CODEC_INDEX_MAX:
       break;
@@ -471,7 +484,7 @@ bool A2dpCodecConfig::setCodecUserConfig(
 
 bool A2dpCodecConfig::codecConfigIsValid(
     const btav_a2dp_codec_config_t& codec_config) {
-  return 
+  return
         (codec_config.codec_type < BTAV_A2DP_CODEC_INDEX_MAX) &&
          (codec_config.sample_rate != BTAV_A2DP_CODEC_SAMPLE_RATE_NONE) &&
          (codec_config.bits_per_sample !=
@@ -1563,10 +1576,29 @@ btav_a2dp_codec_index_t A2DP_SourceCodecIndex(const uint8_t* p_codec_info) {
   return BTAV_A2DP_CODEC_INDEX_MAX;
 }
 
+btav_a2dp_codec_index_t A2DP_SinkCodecIndex(const uint8_t* p_codec_info) {
+  tA2DP_CODEC_TYPE codec_type = A2DP_GetCodecType(p_codec_info);
+
+  LOG_VERBOSE(LOG_TAG, "%s: codec_type = 0x%x", __func__, codec_type);
+
+  switch (codec_type) {
+    case A2DP_MEDIA_CT_SBC:
+      return A2DP_SinkCodecIndexSbc(p_codec_info);
+    case A2DP_MEDIA_CT_AAC:
+      return A2DP_SinkCodecIndexAac(p_codec_info);
+    default:
+      break;
+  }
+   LOG_ERROR(LOG_TAG, "%s: unsupported codec type 0x%x", __func__, codec_type);
+  return BTAV_A2DP_CODEC_INDEX_MAX;
+}
+
 const char* A2DP_CodecIndexStr(btav_a2dp_codec_index_t codec_index) {
   switch (codec_index) {
     case BTAV_A2DP_CODEC_INDEX_SOURCE_SBC:
       return A2DP_CodecIndexStrSbc();
+    case BTAV_A2DP_CODEC_INDEX_SINK_AAC:
+      return A2DP_CodecIndexStrAacSink();
     case BTAV_A2DP_CODEC_INDEX_SINK_SBC:
       return A2DP_CodecIndexStrSbcSink();
     case BTAV_A2DP_CODEC_INDEX_SOURCE_AAC:
@@ -1593,6 +1625,8 @@ bool A2DP_InitCodecConfig(btav_a2dp_codec_index_t codec_index,
   switch (codec_index) {
     case BTAV_A2DP_CODEC_INDEX_SOURCE_SBC:
       return A2DP_InitCodecConfigSbc(p_cfg);
+    case BTAV_A2DP_CODEC_INDEX_SINK_AAC:
+      return A2DP_InitCodecConfigAacSink(p_cfg);
     case BTAV_A2DP_CODEC_INDEX_SINK_SBC:
       return A2DP_InitCodecConfigSbcSink(p_cfg);
     case BTAV_A2DP_CODEC_INDEX_SOURCE_AAC:
@@ -1841,6 +1875,46 @@ bool A2DP_Get_AAC_VBR_Status(const RawAddress *remote_bdaddr) {
    return vbr_supported;
 }
 
+void A2DP_SetSinkCodec(const char *sink_cap){
+  char *tok = NULL;
+  char *tmp_token = NULL;
+
+  tok = strtok_r((char*)sink_cap, "-", &tmp_token);
+  while (tok != NULL)
+  {
+    if (strcmp(tok,"sbc") == 0){
+      LOG_INFO(LOG_TAG,"%s: SBC for Sink is supported",__func__);
+      sbc_sink = true;
+    } else if (strcmp(tok,"aac") == 0) {
+      LOG_INFO(LOG_TAG,"%s: AAC for Sink is supported",__func__);
+      aac_sink = true;
+    }
+    tok = strtok_r(NULL, "-", &tmp_token);
+  }
+  char vbr_value[PROPERTY_VALUE_MAX] = {'\0'};
+  property_get("persist.vendor.qcom.bluetooth.aac_vbr_ctl.enabled", vbr_value, "false");
+  if (!(strcmp(vbr_value,"true"))) {
+    BTIF_TRACE_DEBUG("%s: AAC VBR is enabled", __func__);
+    vbr_supported = true;
+  }
+}
+
+bool A2DP_IsCodecEnabledInSink(btav_a2dp_codec_index_t codec_index) {
+  bool codec_status = false;
+  switch ((int)codec_index) {
+  case BTAV_A2DP_CODEC_INDEX_SINK_SBC:
+    codec_status = sbc_sink;
+    break;
+  case BTAV_A2DP_CODEC_INDEX_SINK_AAC:
+    codec_status = aac_sink;
+    break;
+  case BTAV_A2DP_CODEC_INDEX_SINK_MAX:
+  default:
+    break;
+  }
+  return codec_status;
+}
+
 bool A2DP_GetOffloadStatus() {
   return mA2dp_offload_status;
 }
@@ -1917,6 +1991,7 @@ bool A2DP_IsCodecEnabledInOffload(btav_a2dp_codec_index_t codec_index) {
     case BTAV_A2DP_CODEC_INDEX_SOURCE_APTX_TWS:
       codec_status = aptxtws_offload;
       break;
+    case BTAV_A2DP_CODEC_INDEX_SINK_AAC:
     case BTAV_A2DP_QVA_CODEC_INDEX_SOURCE_MAX:
     case BTAV_A2DP_CODEC_INDEX_SINK_MAX:
     default:
diff --git a/stack/a2dp/a2dp_sbc.cc b/stack/a2dp/a2dp_sbc.cc
index 445c96e56..edd1ae599 100644
--- a/stack/a2dp/a2dp_sbc.cc
+++ b/stack/a2dp/a2dp_sbc.cc
@@ -1029,6 +1029,11 @@ btav_a2dp_codec_index_t A2DP_SourceCodecIndexSbc(
   return BTAV_A2DP_CODEC_INDEX_SOURCE_SBC;
 }
 
+btav_a2dp_codec_index_t A2DP_SinkCodecIndexSbc(
+    UNUSED_ATTR const uint8_t* p_codec_info) {
+  return BTAV_A2DP_CODEC_INDEX_SINK_SBC;
+}
+
 const char* A2DP_CodecIndexStrSbc(void) { return "SBC"; }
 
 const char* A2DP_CodecIndexStrSbcSink(void) { return "SBC SINK"; }
@@ -1781,7 +1786,13 @@ A2dpCodecConfigSbcSink::~A2dpCodecConfigSbcSink() {}
 bool A2dpCodecConfigSbcSink::init() {
   if (!isValid()) return false;
 
-  return true;
+  if (A2DP_IsCodecEnabledInSink(BTAV_A2DP_CODEC_INDEX_SINK_SBC)){
+    LOG_DEBUG(LOG_TAG, "%s: SBC is enabled in Sink", __func__);
+    return true;
+  } else {
+    LOG_DEBUG(LOG_TAG, "%s: SBC is disabled in Sink", __func__);
+    return false;
+  }
 }
 
 bool A2dpCodecConfigSbcSink::useRtpHeaderMarkerBit() const {
diff --git a/stack/a2dp/a2dp_vendor.cc b/stack/a2dp/a2dp_vendor.cc
index baa64ae05..e2887ed29 100644
--- a/stack/a2dp/a2dp_vendor.cc
+++ b/stack/a2dp/a2dp_vendor.cc
@@ -47,6 +47,12 @@
  * limitations under the License.
  */
 
+/*
+ * Changes from Qualcomm Innovation Center are provided under the following license:
+ * Copyright (c) 2022 Qualcomm Innovation Center, Inc. All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
+ */
+
 /**
  * Vendor Specific A2DP Codecs Support
  */
@@ -718,6 +724,7 @@ const char* A2DP_VendorCodecIndexStr(btav_a2dp_codec_index_t codec_index) {
     case BTAV_A2DP_CODEC_INDEX_SOURCE_SBC:
     case BTAV_A2DP_CODEC_INDEX_SINK_SBC:
     case BTAV_A2DP_CODEC_INDEX_SOURCE_AAC:
+    case BTAV_A2DP_CODEC_INDEX_SINK_AAC:
       break;  // These are not vendor-specific codecs
     case BTAV_A2DP_CODEC_INDEX_SOURCE_APTX:
       return A2DP_VendorCodecIndexStrAptx();
@@ -745,6 +752,7 @@ bool A2DP_VendorInitCodecConfig(btav_a2dp_codec_index_t codec_index,
     case BTAV_A2DP_CODEC_INDEX_SOURCE_SBC:
     case BTAV_A2DP_CODEC_INDEX_SINK_SBC:
     case BTAV_A2DP_CODEC_INDEX_SOURCE_AAC:
+    case BTAV_A2DP_CODEC_INDEX_SINK_AAC:
       break;  // These are not vendor-specific codecs
     case BTAV_A2DP_CODEC_INDEX_SOURCE_APTX:
       return A2DP_VendorInitCodecConfigAptx(p_cfg);
diff --git a/stack/avdt/avdt_ccb_act.cc b/stack/avdt/avdt_ccb_act.cc
index ff2bf156d..c6b570c98 100644
--- a/stack/avdt/avdt_ccb_act.cc
+++ b/stack/avdt/avdt_ccb_act.cc
@@ -177,6 +177,8 @@ bool avdt_ccb_check_peer_eligible_for_aac_codec(tAVDT_CCB* p_ccb) {
   uint16_t version = 0;
   bool vndr_prdt_ver_present = false;
   bool aac_support = false;
+  tAVDT_SCB* p_scb = avdt_scb_by_peer_addr(p_ccb->peer_addr);
+
   if (btif_config_get_uint16(p_ccb->peer_addr.ToString().c_str(), PNP_VENDOR_ID_CONFIG_KEY,
       (uint16_t*)&vendor) && btif_config_get_uint16(p_ccb->peer_addr.ToString().c_str(),
       PNP_PRODUCT_ID_CONFIG_KEY, (uint16_t*)&product) &&
@@ -186,7 +188,12 @@ bool avdt_ccb_check_peer_eligible_for_aac_codec(tAVDT_CCB* p_ccb) {
                           vendor, product, version);
     vndr_prdt_ver_present = true;
   }
-  if (vndr_prdt_ver_present && (vendor == A2DP_AAC_BOSE_VENDOR_ID)) {
+
+  if((p_scb != NULL) && (p_scb->cs.tsep == AVDT_TSEP_SNK)
+      &&(p_scb->cs.is_split_enabled)) {
+    AVDT_TRACE_EVENT("%s: Enable AAC for sink role\n", __func__);
+    aac_support = true;
+  } else if (vndr_prdt_ver_present && (vendor == A2DP_AAC_BOSE_VENDOR_ID)) {
     APPL_TRACE_DEBUG("%s: vendor id info matches ", __func__);
     vndr_prdt_ver_present = false;
     aac_support = true;
diff --git a/stack/include/a2dp_aac.h b/stack/include/a2dp_aac.h
index 574b3729e..fec50c749 100644
--- a/stack/include/a2dp_aac.h
+++ b/stack/include/a2dp_aac.h
@@ -14,6 +14,12 @@
  * limitations under the License.
  */
 
+/*
+ * Changes from Qualcomm Innovation Center are provided under the following license:
+ * Copyright (c) 2022 Qualcomm Innovation Center, Inc. All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
+ */
+
 //
 // A2DP Codec API for AAC
 //
@@ -59,6 +65,25 @@ class A2dpCodecConfigAac : public A2dpCodecConfig {
 // to Codec Information Element and returns same CIE.
 bool A2DP_GetAacCIE(const uint8_t* p_codec_info,
                         tA2DP_AAC_CIE *cfg_cie);
+
+class A2dpCodecConfigAacSink : public A2dpCodecConfig {
+ public:
+  A2dpCodecConfigAacSink(btav_a2dp_codec_priority_t codec_priority);
+  virtual ~A2dpCodecConfigAacSink();
+
+              bool init() override;
+  period_ms_t encoderIntervalMs() const override;
+  bool setCodecConfig(const uint8_t* p_peer_codec_info, bool is_capability,
+                      uint8_t* p_result_codec_config) override;
+
+ private:
+  bool useRtpHeaderMarkerBit() const override;
+  bool updateEncoderUserConfig(
+      const tA2DP_ENCODER_INIT_PEER_PARAMS* p_peer_params,
+      bool* p_restart_input, bool* p_restart_output,
+      bool* p_config_updated) override;
+};
+
 // Checks whether the codec capabilities contain a valid A2DP AAC Source
 // codec.
 // NOTE: only codecs that are implemented are considered valid.
@@ -234,11 +259,23 @@ bool A2DP_AdjustCodecAac(uint8_t* p_codec_info);
 // otherwise |BTAV_A2DP_CODEC_INDEX_MAX|.
 btav_a2dp_codec_index_t A2DP_SourceCodecIndexAac(const uint8_t* p_codec_info);
 
+// Gets the A2DP AAC Source codec index for a given |p_codec_info|.
+// Returns the corresponding |btav_a2dp_codec_index_t| on success,
+// otherwise |BTAV_A2DP_CODEC_INDEX_MAX|.
+btav_a2dp_codec_index_t A2DP_SinkCodecIndexAac(const uint8_t* p_codec_info);
+
 // Gets the A2DP AAC Source codec name.
 const char* A2DP_CodecIndexStrAac(void);
 
+// Gets the A2DP AAC Sink codec name.
+const char* A2DP_CodecIndexStrAacSink(void);
+
 // Initializes A2DP AAC Source codec information into |tAVDT_CFG|
 // configuration entry pointed by |p_cfg|.
 bool A2DP_InitCodecConfigAac(tAVDT_CFG* p_cfg);
 
+// Initializes A2DP AAC Sink codec information into |tAVDT_CFG|
+// configuration entry pointed by |p_cfg|.
+bool A2DP_InitCodecConfigAacSink(tAVDT_CFG* p_cfg);
+
 #endif  // A2DP_AAC_H
diff --git a/stack/include/a2dp_codec_api.h b/stack/include/a2dp_codec_api.h
index 0623124a1..54c3c58f9 100644
--- a/stack/include/a2dp_codec_api.h
+++ b/stack/include/a2dp_codec_api.h
@@ -18,6 +18,12 @@
  * limitations under the License.
  */
 
+/*
+ * Changes from Qualcomm Innovation Center are provided under the following license:
+ * Copyright (c) 2022 Qualcomm Innovation Center, Inc. All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
+ */
+
 //
 // A2DP Codecs API
 //
@@ -671,6 +677,11 @@ bool A2DP_AdjustCodec(uint8_t* p_codec_info);
 // otherwise |BTAV_A2DP_CODEC_INDEX_MAX|.
 btav_a2dp_codec_index_t A2DP_SourceCodecIndex(const uint8_t* p_codec_info);
 
+// Gets the A2DP Source codec index for a given |p_codec_info|.
+// Returns the corresponding |btav_a2dp_codec_index_t| on success,
+// otherwise |BTAV_A2DP_CODEC_INDEX_MAX|.
+btav_a2dp_codec_index_t A2DP_SinkCodecIndex(const uint8_t* p_codec_info);
+
 // Gets the A2DP codec name for a given |codec_index|.
 const char* A2DP_CodecIndexStr(btav_a2dp_codec_index_t codec_index);
 
@@ -689,6 +700,8 @@ bool A2DP_IsScramblingSupported();
 bool A2DP_Is44p1kFreqSupported();
 bool A2DP_IsCodecEnabled(btav_a2dp_codec_index_t codec_index);
 bool A2DP_IsCodecEnabledInSoftware(btav_a2dp_codec_index_t codec_index);
+void A2DP_SetSinkCodec(const char *sink_cap);
+bool A2DP_IsCodecEnabledInSink(btav_a2dp_codec_index_t codec_index);
 bool A2DP_Get_AAC_VBR_Status(const RawAddress *remote_bdaddr);
 bool A2DP_Get_Aptx_AdaptiveR2_1_Supported();
 bool A2DP_Get_Aptx_AdaptiveR2_2_Supported();
diff --git a/stack/include/a2dp_sbc.h b/stack/include/a2dp_sbc.h
index 4ac648d3a..dede9f784 100644
--- a/stack/include/a2dp_sbc.h
+++ b/stack/include/a2dp_sbc.h
@@ -262,6 +262,11 @@ bool A2DP_AdjustCodecSbc(uint8_t* p_codec_info);
 // otherwise |BTAV_A2DP_CODEC_INDEX_MAX|.
 btav_a2dp_codec_index_t A2DP_SourceCodecIndexSbc(const uint8_t* p_codec_info);
 
+// Gets the A2DP SBC Source codec index for a given |p_codec_info|.
+// Returns the corresponding |btav_a2dp_codec_index_t| on success,
+// otherwise |BTAV_A2DP_CODEC_INDEX_MAX|.
+btav_a2dp_codec_index_t A2DP_SinkCodecIndexSbc(const uint8_t* p_codec_info);
+
 // Gets the A2DP SBC Source codec name.
 const char* A2DP_CodecIndexStrSbc(void);
 
-- 
2.17.1

